### ISSUE TITLE
Camelcase rule change for #9470

### ISSUE THREAD

**Issue by jbrobst** [eslint/eslint#13041]
**What rule do you want to change?**

camelcase: fix quoted properties being ignored (#9470) but allow use of the old behavior through an option.

**Does this change cause the rule to produce more or fewer warnings?**

More.

**How will the change be implemented? (New option, new default behavior, etc.)?**

The default behavior will be to warn about non-camelcase quoted properties, and the previous behavior of ignoring quoted properties will be behind a new option, `ignoreQuotedProperties`.

**Please provide some example code that this change will affect:**

```js
var foo = {
    barBaz: 1,
    "qux_quux": 2
};
foo.camelCased = 10;
foo["snake_cased"] = 11;
```

**What does the rule currently do for this code?**

Currently, the camelcase rule emits no warnings for this code.

**What will the rule do after it's changed?**

A warning will be produced for the `"qux_quux"` and `"snake_cased"` properties because they are not camelcase. If `ignoreQuotedProperties` is enabled, the old behavior is retained and no warnings are produced.

**Are you willing to submit a pull request to implement this change?**

Yes, see #12777.

**Comment by kaicataldo** [comment]
Can you explain what problem you are trying to solve with this change? It seems to me like the rule is working as intended, and I'd love to have a better understanding of what use cases would benefit from this.

**Comment by jbrobst** [comment]
Since #9470 was accepted as a bug, I'm assuming the consensus for the default behavior of the camelcase rule with `properties: "always"` should be to check all property names regardless of whether they are quoted. Retaining an option for the old behavior is desirable because "quote property names that aren't in camelcase" can be an intentional style. One reason for this is that there are instances that require the use of property names that aren't in camelcase. An example is the use of the terser API, which requires the use of an object with snake case properties. Another example, which occurs in eslint's own tests, is using file paths as property names, which don't make sense to validate as camelcase. Users like myself or the authors of those eslint tests are currently leveraging the quoting behavior as if it is an intended part of the rule; `ignoreQuotedProperties` makes this style a deliberate choice.

Of course, if by "the rule is working as intended" you mean that "don't check quoted properties" is intentional and #9470 is not a bug, then neither this rule change nor my PR is necessary. If that is the case, I think a mention that `properties: "always"` does not validate quoted properties should be added to the camelcase rule's documentation.

**Comment by kaicataldo** [comment]
Understood, thank you. I misunderstood the original proposal - sorry about that! I think this makes sense and I support it, though I think we should put this check behind an option so that it isn't a breaking change.

**Comment by jbrobst** [comment]
If we have `ignoreQuotedProperties` default to true instead of false this will be non-breaking, what do you think?

**Comment by mdjermanovic** [comment]
> If we have `ignoreQuotedProperties` default to true instead of false this will be non-breaking, what do you think?

I think the default value should be non-breaking `true`.

The actual behavior of this rule may be an oversight, but it has some use cases (e.g., to quote names required by a third-party API), so it might be better to retain it for the start and maybe change the default later in another major version.

**Comment by jbrobst** [comment]
Sounds good; I've updated the PR with `ignoreQuotedProperties: true` as the default and with 0c20bc068e608869981a10711bba88ffde1539d8 incorporated.
